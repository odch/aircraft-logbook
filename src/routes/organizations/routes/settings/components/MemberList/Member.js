import React from 'react'
import PropTypes from 'prop-types'
import { injectIntl } from 'react-intl'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction'
import IconButton from '@material-ui/core/IconButton'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import Tooltip from '@material-ui/core/Tooltip'
import { withStyles } from '@material-ui/core'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import {
  member as memberShape,
  intl as intlShape
} from '../../../../../../shapes'
import { formatDate, formatTime } from '../../../../../../util/dates'

const styles = {
  pending: {
    fontStyle: 'italic',
    color: '#AAAAAA'
  }
}

class Member extends React.Component {
  state = {
    anchorEl: null
  }

  handleMenuOpen = event => {
    this.setState({ anchorEl: event.currentTarget })
  }

  handleMenuClose = () => {
    this.setState({ anchorEl: null })
  }

  handleMenuItemClick = (callback, args) => () => {
    if (typeof callback === 'function') {
      callback(...args)
    }
    this.handleMenuClose()
  }

  getRoleNames = roles =>
    roles
      ? roles.map(role => this.msg(`organization.role.${role}`)).join(', ')
      : ''

  msg = (id, values) => this.props.intl.formatMessage({ id }, values)

  getPendingInvitationTooltip = member => {
    if (member.inviteTimestamp) {
      return this.msg('organization.invite.pending', {
        inviteDate: formatDate(member.inviteTimestamp),
        inviteTime: formatTime(member.inviteTimestamp)
      })
    }
    return this.msg('organization.invite.notsent')
  }

  render() {
    const { member, classes } = this.props

    const invitePending = !member.user

    const itemText = (
      <ListItemText
        primary={`${member.lastname} ${member.firstname}${
          member.nr ? ` (${member.nr})` : ''
        }`}
        secondary={this.getRoleNames(member.roles)}
      />
    )

    return (
      <React.Fragment>
        <ListItem
          key={member.id}
          disableGutters
          className={invitePending ? classes.pending : undefined}
          divider
        >
          {invitePending ? (
            <Tooltip
              title={this.getPendingInvitationTooltip(member)}
              placement="bottom-start"
            >
              {itemText}
            </Tooltip>
          ) : (
            itemText
          )}
          <ListItemSecondaryAction>
            <IconButton onClick={this.handleMenuOpen}>
              <MoreVertIcon />
            </IconButton>
          </ListItemSecondaryAction>
        </ListItem>
        {this.state.anchorEl && this.renderMenu()}
      </React.Fragment>
    )
  }

  renderMenu() {
    const { member, openEditMemberDialog, openDeleteMemberDialog } = this.props

    return (
      <Menu anchorEl={this.state.anchorEl} onClose={this.handleMenuClose} open>
        <MenuItem
          onClick={this.handleMenuItemClick(openEditMemberDialog, [member])}
        >
          {this.msg('organization.settings.member.edit')}
        </MenuItem>
        <MenuItem
          onClick={this.handleMenuItemClick(openDeleteMemberDialog, [member])}
        >
          {this.msg('organization.settings.member.delete')}
        </MenuItem>
      </Menu>
    )
  }
}

Member.propTypes = {
  member: memberShape.isRequired,
  openDeleteMemberDialog: PropTypes.func.isRequired,
  openEditMemberDialog: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
  classes: PropTypes.object.isRequired
}

export default injectIntl(withStyles(styles)(Member))
