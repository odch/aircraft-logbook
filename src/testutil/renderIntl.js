import React from 'react'
import renderer from 'react-test-renderer'
import { createMount } from '@material-ui/core/test-utils'
import { IntlProvider } from 'react-intl'
import pretty from 'pretty'
import messages from '../messages'

const LOCALE = 'de'

const renderInIntlProvider = component => (
  <IntlProvider locale={LOCALE} messages={messages[LOCALE]}>
    {component}
  </IntlProvider>
)

const renderIntl = (component, options) =>
  renderer.create(renderInIntlProvider(component), options)

export const renderIntlUpdate = (renderer, component, options) =>
  renderer.update(renderInIntlProvider(component), options)

export const renderIntlMaterial = (component, returnPrettyHtml = false) => {
  const rendered = createMount()(renderInIntlProvider(component))
  return returnPrettyHtml ? pretty(rendered.html()) : rendered
}

export default renderIntl
