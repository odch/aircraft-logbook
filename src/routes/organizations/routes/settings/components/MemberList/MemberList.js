import React from 'react'
import PropTypes from 'prop-types'
import { injectIntl } from 'react-intl'
import List from '@material-ui/core/List'
import TablePagination from '@material-ui/core/TablePagination'
import {
  member as memberShape,
  intl as intlShape
} from '../../../../../../shapes'
import isLoaded from '../../../../../../util/isLoaded'
import LoadingIcon from '../../../../../../components/LoadingIcon'
import Member from './Member'
import DeleteMemberDialog from '../DeleteMemberDialog'

class MemberList extends React.Component {
  componentDidMount() {
    const { organizationId, fetchMembers } = this.props
    fetchMembers(organizationId)
  }

  msg = id => this.props.intl.formatMessage({ id })

  render() {
    const {
      organizationId,
      members,
      pagination,
      deleteMemberDialog,
      openDeleteMemberDialog,
      closeDeleteMemberDialog,
      deleteMember,
      setMembersPage
    } = this.props

    if (!isLoaded(members)) {
      return <LoadingIcon />
    }

    return (
      <React.Fragment>
        <List>
          {members.map(member => (
            <Member
              key={member.id}
              member={member}
              openDeleteMemberDialog={openDeleteMemberDialog}
            />
          ))}
        </List>
        <TablePagination
          component="div"
          count={pagination.rowsCount}
          rowsPerPage={pagination.rowsPerPage}
          rowsPerPageOptions={[]}
          page={pagination.page}
          onChangePage={(event, page) => setMembersPage(page)}
        />
        {deleteMemberDialog && deleteMemberDialog.open && (
          <DeleteMemberDialog
            organizationId={organizationId}
            member={deleteMemberDialog.member}
            submitting={deleteMemberDialog.submitting}
            onConfirm={() =>
              deleteMember(organizationId, deleteMemberDialog.member.id)
            }
            onClose={closeDeleteMemberDialog}
          />
        )}
      </React.Fragment>
    )
  }
}

MemberList.propTypes = {
  organizationId: PropTypes.string.isRequired,
  members: PropTypes.arrayOf(memberShape),
  pagination: PropTypes.shape({
    rowsCount: PropTypes.number.isRequired,
    page: PropTypes.number.isRequired,
    rowsPerPage: PropTypes.number.isRequired
  }),
  deleteMemberDialog: PropTypes.shape({
    open: PropTypes.bool,
    submitting: PropTypes.bool,
    member: memberShape
  }),
  fetchMembers: PropTypes.func.isRequired,
  openDeleteMemberDialog: PropTypes.func.isRequired,
  closeDeleteMemberDialog: PropTypes.func.isRequired,
  deleteMember: PropTypes.func.isRequired,
  setMembersPage: PropTypes.func.isRequired,
  intl: intlShape.isRequired
}

export default injectIntl(MemberList)
