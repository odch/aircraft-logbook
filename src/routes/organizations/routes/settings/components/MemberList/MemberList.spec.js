import React from 'react'
import renderIntl from '../../../../../../testutil/renderIntl'
import MemberList from './MemberList'

describe('routes', () => {
  describe('organizations', () => {
    describe('routes', () => {
      describe('settings', () => {
        describe('components', () => {
          describe('MemberList', () => {
            it('calls fetchMembers when mounted', () => {
              const fetchMembers = jest.fn()

              renderIntl(
                <MemberList
                  organizationId="my_org"
                  members={undefined}
                  fetchMembers={fetchMembers}
                  openDeleteMemberDialog={() => {}}
                  closeDeleteMemberDialog={() => {}}
                  deleteMember={() => {}}
                  setMembersPage={() => {}}
                />
              )

              expect(fetchMembers).toBeCalledWith('my_org')
            })

            it('renders loading icon if members not loaded', () => {
              const tree = renderIntl(
                <MemberList
                  organizationId="my_org"
                  members={undefined}
                  fetchMembers={() => {}}
                  openDeleteMemberDialog={() => {}}
                  closeDeleteMemberDialog={() => {}}
                  deleteMember={() => {}}
                  setMembersPage={() => {}}
                />
              ).toJSON()
              expect(tree).toMatchSnapshot()
            })

            it('renders members if loaded', () => {
              const tree = renderIntl(
                <MemberList
                  organizationId="my_org"
                  members={[
                    {
                      id: 'member1',
                      firstname: 'Hans',
                      lastname: 'Keller',
                      nr: '34646',
                      roles: ['manager']
                    },
                    {
                      id: 'member2',
                      firstname: 'Max',
                      lastname: 'Muster',
                      roles: ['user']
                    }
                  ]}
                  deleteMemberDialog={{ open: false }}
                  pagination={{
                    rowsCount: 2,
                    page: 0,
                    rowsPerPage: 10
                  }}
                  fetchMembers={() => {}}
                  openDeleteMemberDialog={() => {}}
                  closeDeleteMemberDialog={() => {}}
                  deleteMember={() => {}}
                  setMembersPage={() => {}}
                />
              ).toJSON()
              expect(tree).toMatchSnapshot()
            })
          })
        })
      })
    })
  })
})
