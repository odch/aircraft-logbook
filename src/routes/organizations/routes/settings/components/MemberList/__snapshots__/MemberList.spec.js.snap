// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes organizations routes settings components MemberList renders loading icon if members not loaded 1`] = `
<div
  className="LoadingIcon-container-1"
>
  <div
    className="MuiCircularProgress-root LoadingIcon-icon-2 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;

exports[`routes organizations routes settings components MemberList renders members if loaded 1`] = `
Array [
  <ul
    className="MuiList-root MuiList-padding"
  >
    <li
      className="MuiListItem-container"
    >
      <div
        className="MuiListItem-root MuiListItem-divider MuiListItem-secondaryAction"
        disabled={false}
      >
        <div
          className="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Keller Hans (34646)
          </span>
          <p
            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Manager
          </p>
        </div>
      </div>
      <div
        className="MuiListItemSecondaryAction-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </li>
    <li
      className="MuiListItem-container"
    >
      <div
        className="MuiListItem-root Member-pending-19 MuiListItem-divider MuiListItem-secondaryAction"
        disabled={false}
      >
        <div
          aria-describedby={null}
          className="MuiListItemText-root MuiListItemText-multiline"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="Noch nicht beigetreten"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Muster Max
          </span>
          <p
            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Benutzer
          </p>
        </div>
      </div>
      <div
        className="MuiListItemSecondaryAction-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </li>
  </ul>,
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <span
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-caption MuiTypography-colorInherit"
      >
        1-2 of 2
      </span>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
